src/app/root-store/user-tags/user-tags.reducer.ts

```diff
@@ -0,0 +1,55 @@
+import { UserTagsActionsUnion, UserTagsActionTypes } from './user-tags.actions';
+import { Tag } from '../../shared/models';
+import { initialState, UserTagsState } from './user-tags.state';
+
+
+function getEntities(entities: Tag[]): { [id: string]: Tag } {
+  return entities.reduce((dictionary, entity) => {
+    dictionary[entity.key] = entity;
+    return dictionary;
+  }, {})
+}
+
+function getIds(entities: Tag[]): string[] {
+  return entities.map(tag => tag.key)
+}
+
+function upsertMany(entities: Tag[], state: UserTagsState): UserTagsState {
+  const upsertIds = getIds(entities);
+  const upsertEntities = getEntities(entities);
+  return {
+    ...state,
+    ids: [...state.ids, ...upsertIds],
```

Comment:
could there be some duplicate ids?

Reply:
You right, it could be. Changed with ngrx/entities.